### PR TITLE
*: Speed up the add columns phase

### LINKED ISSUE #2729: Speed up add column phase

### PR DESCRIPTION
**Pull request by zimulala**
Fix #2729

### CODE REVIEW

File: model/model.go [hunk 1]
```diff
@@ -58,13 +58,14 @@ func (s SchemaState) String() string {
...
-	Name            CIStr       `json:"name"`
-	Offset          int         `json:"offset"`
-	DefaultValue    interface{} `json:"default"`
-	types.FieldType `json:"type"`
-	State           SchemaState `json:"state"`
-	Comment         string      `json:"comment"`
+	ID                 int64       `json:"id"`
+	Name               CIStr       `json:"name"`
+	Offset             int         `json:"offset"`
+	OriginDefaultValue interface{} `json:"origin default"`
```
Comment (coocood) [inline model/model.go]: origin_default

File: ddl/column.go [hunk 2]
```diff
@@ -165,7 +165,7 @@ func (d *ddl) onAddColumn(t *meta.Meta, job *model.Job) error {
 		if err != nil {
 			return errors.Trace(err)
 		}
-		if columnInfo.DefaultValue != nil || mysql.HasNotNullFlag(columnInfo.Flag) {
+		if columnInfo.DefaultValue == nil && mysql.HasNotNullFlag(columnInfo.Flag) {
```
Comment (coocood) [inline ddl/column.go]: In this case, the column value will be filled with zero value.

**Comment by zimulala** [comment]
PTAL @coocood @shenli

File: ddl/column.go [hunk 3]
```diff
@@ -294,24 +278,10 @@ func (d *ddl) addTableColumn(t table.Table, columnInfo *model.ColumnInfo, reorgI
...
-			job.State = model.JobCancelled
-			log.Errorf("[ddl] fatal: this case shouldn't happen, column %v err %v", columnInfo, err)
-			return errors.Trace(err)
-		}
-	} else if mysql.HasNotNullFlag(columnInfo.Flag) {
-		colMeta.defaultVal = table.GetZeroValue(columnInfo)
-	}
-	for _, col := range t.Meta().Columns {
-		colMeta.oldColMap[col.ID] = &col.FieldType
+	colMeta, err := newColumnMeta(ctx, t, columnInfo, job)
```
Comment (coocood) [inline ddl/column.go]: We don't need to change it if `addTableColumn` is never called.

File: ddl/column.go [hunk 4]
```diff
@@ -398,6 +368,31 @@ type columnMeta struct {
 	oldColMap  map[int64]*types.FieldType
 }
 
+func newColumnMeta(ctx context.Context, t table.Table, columnInfo *model.ColumnInfo, job *model.Job) (*columnMeta, error) {
```
Comment (coocood) [inline ddl/column.go]: It is not needed anymore.

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -798,6 +798,10 @@ func (d *ddl) AddColumn(ctx context.Context, ti ast.Ident, spec *ast.AlterTableS
 	if err != nil {
 		return errors.Trace(err)
 	}
+	if col.DefaultValue == nil && mysql.HasNotNullFlag(col.Flag) {
+		col.DefaultValue = table.GetZeroValue(col.ToInfo())
+	}
+	col.OriginDefaultValue = col.DefaultValue
```
Comment (coocood) [inline ddl/ddl_api.go]: Should make sure originDefaultValue is not a time function.
OriginDefaultValue should be either string or nil.

Comment (zimulala) [inline ddl/ddl_api.go]: It is done in line 797.

File: executor/distsql.go [hunk 6]
```diff
@@ -942,17 +942,25 @@ func statementContextToFlags(sc *variable.StatementContext) uint64 {
 
 func setPBColumnsDefaultValue(ctx context.Context, pbColumns []*tipb.ColumnInfo, columns []*model.ColumnInfo) error {
 	for i, c := range columns {
-		if c.State == model.StatePublic {
+		if c.State == model.StatePublic && c.OriginDefaultValue == nil {
```
Comment (zimulala) [inline executor/distsql.go]: If `c.State != model.StatePublic` is true, we shoudn't do the following logic.

**Comment by zimulala** [comment]
PTAL @coocood

File: executor/ddl_test.go [hunk 7]
```diff
@@ -145,7 +147,12 @@ func (s *testSuite) TestAlterTableAddColumn(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
 	tk.MustExec("use test")
 	tk.MustExec("create table if not exists alter_test (c1 int)")
-	tk.MustExec("alter table alter_test add column c2 int")
+	tk.MustExec("insert into alter_test values(1)")
+	now := time.Now().Format(types.TimeFormat)
+	tk.MustExec("alter table alter_test add column c2 timestamp default current_timestamp")
+	time.Sleep(1 * time.Second)
```
Comment (coocood) [inline executor/ddl_test.go]: Sleep one second in test is expensive, can we remove it?

File: executor/ddl_test.go [hunk 8]
```diff
@@ -145,7 +147,12 @@ func (s *testSuite) TestAlterTableAddColumn(c *C) {
...
+	rowStr := fmt.Sprintf("%v", now)
+	tk.MustQuery("select c2 from alter_test").Check(testkit.Rows(rowStr))
```
Comment (coocood) [inline executor/ddl_test.go]: The test may fail if now is at the end of a second.

File: ddl/ddl_api.go [hunk 9]
```diff
@@ -798,6 +800,13 @@ func (d *ddl) AddColumn(ctx context.Context, ti ast.Ident, spec *ast.AlterTableS
...
+	if col.OriginDefaultValue == expression.CurrentTimestamp {
```
Comment (coocood) [inline ddl/ddl_api.go]: Need to check type is datetime or timestamp.

File: executor/ddl_test.go (hunk 8 above)
Comment (zimulala) [inline executor/ddl_test.go]: Yep. But it used the local store, its lease is 0.

File: executor/ddl_test.go (hunk 7 above)
Comment (zimulala) [inline executor/ddl_test.go]: This is used to check if the time is created.

File: table/tables/tables.go [hunk 10]
```diff
@@ -462,10 +462,22 @@ func (t *Table) RowWithCols(ctx context.Context, h int64, cols []*table.Column)
 			continue
 		}
 		ri, ok := row[col.ID]
-		if !ok && mysql.HasNotNullFlag(col.Flag) {
+		if ok {
+			v[i] = ri
+			continue
+		}
+
+		if col.OriginDefaultValue != nil && col.State == model.StatePublic {
```
Comment (coocood) [inline table/tables/tables.go]: Only need to check `col.OriginDefaultValue != nil`

Comment (zimulala) [inline table/tables/tables.go]: Read operation needs the column state is public.

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@shenli PTAL

File: ddl/ddl_api.go [hunk 11]
```diff
@@ -282,7 +283,6 @@ func columnDefToCol(ctx context.Context, offset int, colDef *ast.ColumnDef) (*ta
 					return nil, nil, ErrInvalidOnUpdate.Gen("invalid ON UPDATE for - %s", col.Name)
 				}
 
-				col.Flag |= mysql.OnUpdateNowFlag
```
Comment (shenli) [inline ddl/ddl_api.go]: Why remove this?

**Comment by shenli** [comment]
LGTM

**Pull request merged**